### PR TITLE
[exporter/tanzuobservability] Make metrics stanza optional in config.

### PR DESCRIPTION
**Pull request by keep94**
**Description:** 
Fixed bug.  The collector failed to start if the tanzuobservability exporter was configured with traces only and not metrics and the traces endpoint had anything other than localhost. This is a bug because traces came first, and metrics came later. If a user has only traces configured for tanzuobservability exporter, they should be able to start the collector even if they have upgraded to a version that supports metrics too.

To fix, I made the Metrics field in our config struct be a pointer field instead of a struct field, so that it will be nil if it is missing.  If metrics aren't configured in the config file, we don't send any metrics.

**Link to tracking Issue:** Not available to public.

**Testing:**. 
1. Unit testing.
2. Test that I can start the collector when the traces endpoint is http://192.168.XX.XX:30001 and metrics are not configured and that traces are actually sent.
3. Test that metrics get sent when the metrics endpoint is set to http://192.168.XX.XX:2878.

**Documentation:** Changes to README.md

### CODE REVIEW

File: exporter/tanzuobservabilityexporter/metrics_exporter.go
Comment (oppegard) [inline exporter/tanzuobservabilityexporter/metrics_exporter.go:64]: See my comment in `exporter.go` regarding using the config's struct values directly.

**Review by oppegard — CHANGES_REQUESTED** [review]

File: exporter/tanzuobservabilityexporter/exporter.go
Comment (keep94) [reply exporter/tanzuobservabilityexporter/exporter.go:70]: The missing URLs do get filled in the returned config struct.  In fact, you could run url.Parse(cfg.Traces.Endpoint). However, withoutMissingFields() has to call url.Parse already to do its job of filling in missing fields in the struct.  Rather than make the caller parse the URL strings again, why not just return the url.URL values?  If you prefer, I can change withoutMissingFields() to return just the config struct and the error, but keep in mind that if we do that, the caller will have to parse the URL strings, check for errors, etc.

**Review by oppegard — APPROVED** [review]

**Comment by keep94** [comment]
It appears that unit tests in the exporter/prometheusremotewriteexporter are failing, which is a different exporter than what this PR is for.  What can be done to resolve?

**Comment by jpkrohling** [comment]
Failed tests restarted.

**Comment by keep94** [comment]
Even though the tests were restarted it looks like: `github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusremotewriteexporter` is still failing.  I am not sure why.  If possible, we would like to have this merged before the next big release because it fixes a severe backward compatibility bug with the config file.

**Comment by keep94** [comment]
Hi. Thank you for restarting the tests.  It looks like the build for this PR is now green.  Are there any upcoming comments for this PR?  As a gentle reminder we want to get this in a release as soon as possible because it fixes a major backward compatibility bug.  Thank you.

**Comment by keep94** [comment]
Hello.  This code has been approved my my colleagues and the build is green.  Is anything else needed for this code to be reviewed?  This PR contains a fix for a backward compatibility bug in the configuration for this exporter.  We would like it to be merged and released sooner than later.  Thank you for your time. Hope to hear from you soon.

**Comment by keep94** [comment]
@codeboten Hello. This PR, #9098, still needs to be reviewed by an owner.  I notice that your name is on this PR to review it.  Is there anything you need from us to facilitate review of this code?  As a gentle reminder, this PR fixes a major bug related to backward compatibility issues in the config file; therefore, we would like this to be merged to the main branch as soon as possible.  Thank you for your attention.

File: CHANGELOG.md
Comment (pmm-sumo) [inline CHANGELOG.md:38]: Something is wrong after the rebase. Also, the fix description should go to around line 16 or so

File: exporter/tanzuobservabilityexporter/config.go
Comment (pmm-sumo) [inline exporter/tanzuobservabilityexporter/config.go:129]: I am not confident this is good use of panic. Maybe someone more experienced with Go could confirm. I was anticipating returning the error instead

File: CHANGELOG.md
Comment (keep94) [reply CHANGELOG.md]: I rebased and pulled the tanzuobservability exporter change up into the unreleased section.

File: exporter/tanzuobservabilityexporter/config.go
Comment (keep94) [reply exporter/tanzuobservabilityexporter/config.go]: I got rid of the panic and return the error instead.

File: exporter/tanzuobservabilityexporter/config.go
Comment (TylerHelmuth) [inline exporter/tanzuobservabilityexporter/config.go:76]: I'm a little confused.  If the user does not configure metrics, this is adding metrics configuration and using the supplied trace endpoint right? I think the part I'm missing is why the metrics settings need filled out in order for the exporter to work if it is only defined in a trace pipeline?  Or is the issue that users are adding the exporter to the metrics pipeline but not supplying metrics configuration?

Comment (pmm-sumo) [reply exporter/tanzuobservabilityexporter/config.go:76]: Good point. I think the intention (please correct me @keep94) was to be able to separately specify metrics configuration and traces configuration and when only one is available, assume it's like the other but on a different (default) port. 

Perhaps that's something tied to tanzuobservability but if not, I would suggest to consider approach [similar to used in OTLP/HTTP](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/otlphttpexporter#otlphttp-exporter). Either a single endpoint can be configured (in this case, this would be the hostname) or specifically `traces_endpoint/metrics_endpoint/logs_endpoint`. 

Also, if endpoint for the signal was not defined and it's used in the pipeline it should just trigger an error

Comment (keep94) [reply exporter/tanzuobservabilityexporter/config.go:76]: I haven't verified this yet, but perhaps the `createMetricsExporter` function only gets called when our exporter is used in a metrics pipeline. Perhaps all I have to do is throw an error in `createMetricsExporter` if there is no metrics configuration. But, I just remembered that for tanzuobservability we have a unique requirement.  Our traces exporter uses both the metrics and traces endpoint.  Before we added the metrics exporter, we hardcoded the metrics endpoint to be the same as the traces endpoint except use port 2878.  Now that we have both the metrics and traces exporter, we want the traces exporter to use the same metrics endpoint as the metrics exporter

Comment (pmm-sumo) [reply exporter/tanzuobservabilityexporter/config.go:76]: > Initially I wanted to throw an error if our tanzuobservability exporter is used in the metrics pipeline but no metrics configuration is defined, but I wasn't able to figure out how to do this in the code.

Unfortunately, `Validate()` can be only used in context of Config, without awareness of which pipeline it is. You might still throw an error when e.g. `createTraces` is called and not all config is available. 

Another option is that you might prefer hostname and port numbers. The port numbers would have default values, so they would always point somewhere. This has the limitation of not supporting url's with paths.

Looking at the ideas, I would circle back to how otlphttp exporter does it. Perhaps you could have three fields:

- `hostname` - if set, default port numbers are used and you can support both metrics and traces
- `metrics_endpoint` - by default: "https://\<hostname\>:2878" but could be overwritten here if needed
- `traces_endpoint` - be default: "https://\<hostname\>:30001" but could be overwritten here if needed

So a valid configuration would be:
```yaml
hostname: myserver.example.com
```

or:
```yaml
hostname: myserver.example.com
metrics_endpoint: https://myotherserver.example.com:2878
```

or:
```yaml
traces_endpoint: https://myserver1.example.com:30001
metrics_endpoint: https://myotherserver.example.com:2878
```

Comment (TylerHelmuth) [reply exporter/tanzuobservabilityexporter/config.go:76]: I think this idea a lot.  I think it will simplify a lot of config.go, since you can be sure there is always an endpoint to use.

Comment (oppegard) [reply exporter/tanzuobservabilityexporter/config.go:76]: Hi @pmm-sumo and @TylerHelmuth: I'm a maintainer of this exporter (in addition to @keep94), and I had question. I'll let Travis consider your suggestions and reply on his own.

Your config changes are backwards-incompatible with our existing yaml. Is there an otel-collector procedure or policy for introducing breaking changes like this? Whether or not we go down that path in this PR, it'd be good to know how to go about breaking changes in the future. I'm pretty sure I've seen other exporters change or deprecate config options, so I do think it's possible. Thanks!

Comment (pmm-sumo) [reply exporter/tanzuobservabilityexporter/config.go:76]: Hi @oppegard, I think this config change could be done in a non-breaking way. E.g. the new configuration fields could be added. When user uses the old one (essentially "endpoint"), a deprecation warning shows up in the logs.
After some time, the old configuration fields can be removed then.

Ultimately, I think it's your call since this is your exporter :)

Comment (TylerHelmuth) [reply exporter/tanzuobservabilityexporter/config.go:76]: Example of deprecated config: 

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b9915f749b7ec03fabcff3200bc659e3fe9f549d/processor/cumulativetodeltaprocessor/config.go#L35

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b9915f749b7ec03fabcff3200bc659e3fe9f549d/processor/cumulativetodeltaprocessor/processor.go#L111

Agree with @pmm-sumo, its your exporter so you can handle this however you want.

**Comment by TylerHelmuth** [comment]
Just did a quick review of this and so far I find this iteration easier to interpret.  If I am reading things correctly, both a trace endpoint and metrics endpoint can be supplied, but they must have the same hostname.  Trace endpoint must be supplied, but metrics endpoint is optional.  If metrics endpoint is not supplied, a default port is used with the trace endpoint.  If a metrics endpoint is supplied then the port on the supplied endpoint is used as the metrics port.

My only question right now is did you mean to remove your updates to the README.md?  That was a beautiful README file with some great detail.

**Comment by keep94** [comment]
Hello. Based on what was discussed, I now have brand new code. I will reach out again once my team has had a chance to approve the PR.

**Comment by keep94** [comment]
@TylerHelmuth Thanks for pointing out the missing README.md, I will work on getting those changes back in this PR.  As for your comment, both the traces and metrics endpoints are optional, neither have to be supplied, but if there is a traces pipeline, the traces endpoint must be supplied. If there is a metrics pipeline, the metrics endpoint must be supplied.

**Review by oppegard — APPROVED** [review]

File: exporter/tanzuobservabilityexporter/config.go
Comment (oppegard) [reply exporter/tanzuobservabilityexporter/config.go]: Thanks @pmm-sumo and @TylerHelmuth!

**Comment by keep94** [comment]
Hello everyone. My tech lead approved these changes yesterday.  You may resume reviewing this PR.  Thank you.

**Review by pmm-sumo — COMMENTED** [review]
I think it looks OK, need a rebasing though (we have released v0.50.0 in the meantime and this impacts changelog)

**Comment by TylerHelmuth** [comment]
> @TylerHelmuth Thanks for pointing out the missing README.md, I will work on getting those changes back in this PR. As for your comment, both the traces and metrics endpoints are optional, neither have to be supplied, but if there is a traces pipeline, the traces endpoint must be supplied. If there is a metrics pipeline, the metrics endpoint must be supplied.

I either missed the`metrics_exporter.go` file or I took a look too soon.  Looked over things again today and I continue to like this iteration more than the original.

**Review by TylerHelmuth — APPROVED** [review]

**Comment by keep94** [comment]
Hello. I just rebased this PR to resolve the conflicts in CHANGELOG.md

**Comment by keep94** [comment]
It looks like CHANGELOG.md became out of data even since this morning when I rebased.  Is there a way we can coordinate a time for when I should rebase and then somebody can merge immediately after I rebase?

**Comment by keep94** [comment]
Just rebased to resolve conflicts with CHANGELOG.md

**Comment by keep94** [comment]
Looks like this PR can still be merge automatically. There are failing load tests, but I believe those are flaky because nothing changed except resolving conflicts in CHANGELOG.md.

**Review by codeboten — COMMENTED** [review]
@pmm-sumo please approve if your comments have been addressed 👍

File: exporter/tanzuobservabilityexporter/config.go
Comment (pmm-sumo) [inline exporter/tanzuobservabilityexporter/config.go:100]: I am not sure about named returns in go, i.e. perhaps it should more explicit like following. Up to you @keep94 

```suggestion
	return hostName, port, nil
```

File: exporter/tanzuobservabilityexporter/exporter.go
Comment (pmm-sumo) [inline exporter/tanzuobservabilityexporter/exporter.go:76]: A nit, I would put any "magic value" as a const

**Review by pmm-sumo — APPROVED** [review]
Some nits that are not blocking from my end

File: exporter/tanzuobservabilityexporter/config.go
Comment (keep94) [reply exporter/tanzuobservabilityexporter/config.go]: Done.

**Comment by pmm-sumo** [comment]
> It looks like CHANGELOG.md became out of data even since this morning when I rebased. Is there a way we can coordinate a time for when I should rebase and then somebody can merge immediately after I rebase?

I think that such conflicts are actually trivial to resolve and I think usually this is done by one of the maintainers when merging the PR. So no need to worry about these :)

**Review by codeboten — APPROVED** [review]

**Pull request merged**
